File: py_opencv/masking.py
import cv2 as cv
import numpy as np


img = cv.imread("photos/dog1.jpg")
img = cv.resize(img, (int(img.shape[1] * 0.2), int(img.shape[0]*0.2)), interpolation=cv.INTER_LINEAR)
cv.imshow("dog",img)

blank = np.zeros(img.shape[:2], dtype='uint8')
cv.imshow('blank', blank)
circle_mask = cv.circle(blank.copy(), (img.shape[1]//2, img.shape[0]//2 + 45), 100, 255, -1)
cv.imshow('mask', circle_mask)
rectangle = cv.rectangle(blank.copy(), (300, 300),(400, 400), 255, -1)
weird_shape = cv.bitwise_and(rectangle, circle_mask)


masked_img = cv.bitwise_and(img, img, mask=circle_mask)
cv.imshow("masked img", masked_img)

weirdshape_img = cv.bitwise_and(img, img, mask=weird_shape)
cv.imshow("weird shape", weirdshape_img)

cv.waitKey(0)

File: cpp_opencv/masking/masking.cpp
#include <opencv2/opencv.hpp>
#include <vector>
#include <stdio.h>

using namespace cv;

std::string photo = "../../photos/dog1.jpg";

int main(){
    Mat image = imread(photo);
    resize(image, image, cv::Size(), 0.2, 0.2);
    imshow("dog", image);

    // blank canvas
    Mat blank = Mat::zeros(image.rows, image.cols, CV_8U);
    imshow("blank", blank);

    Mat circle_mask = blank.clone();
    circle(circle_mask, Point(image.cols/2, image.rows/2), 100, Scalar(255,255,255), -1, 8, 0);
    imshow("circle", circle_mask);

    // bitwise and
    Mat masked_photo; 
    bitwise_and(image, image, masked_photo, circle_mask);
    imshow("masked_photo", masked_photo);

    waitKey(0);
    return 0;
}